certora/specs/ERC20.spec

```diff
@@ -0,0 +1,252 @@
+    function balanceOf(address)                    external returns (uint256) envfree;
+    function allowance(address,address)            external returns (uint256) envfree;
+    function owner()                               external returns address   envfree;
+    function delegatedVotingPower(address)         external returns uint256   envfree;
+    function delegatee(address)                    external returns address   envfree;
+    function nonces(address)                       external returns (uint256) envfree;
+    function DOMAIN_SEPARATOR()                    external returns (bytes32) envfree;
+    function upgradeToAndCall(address, bytes memory) internal => CONSTANT;
+}
+/*
+┌─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
+│ Ghost & hooks: sum of all balances                                                                                  │
+└─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
+*/
+ghost mathint sumOfBalances {
+    init_state axiom sumOfBalances == 0;
+}
+
+// Slot is ERC20Storage._balances slot
+hook Sload uint256 balance (slot 0x52c63247e1f47db19d5ce0460030c497f067ca4cebf71ba98eeadabe20bace00)[KEY address addr] {
+   require sumOfBalances >= to_mathint(balance);
+}
+
+// Slot is ERC20Storage._balances slot
+hook Sstore (slot 0x52c63247e1f47db19d5ce0460030c497f067ca4cebf71ba98eeadabe20bace00)[KEY address addr] uint256 newValue (uint256 oldValue) {
+   sumOfBalances = sumOfBalances - oldValue + newValue;
+}
+
+/*
+┌─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
+│ Invariant: totalSupply is the sum of all balances                                                                   │
+└─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
+*/
+
+invariant totalSupplyIsSumOfBalances()
+    to_mathint(totalSupply()) == sumOfBalances;
+
+
+/*
+┌─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
+│ Invariant: balance of address(0) is 0                                                                               │
+└─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
+*/
+
+invariant zeroAddressNoBalance()
+    balanceOf(0) == 0;
+
+/*
+┌─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
+│ Rules: only the token holder (or a permit) can increase allowance. The spender can decrease it by using it          │
+└─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
+*/
+rule onlyHolderOfSpenderCanChangeAllowance(env e, method f){
```

Comment:
```suggestion
rule onlyHolderOrSpenderCanChangeAllowance(env e, method f){
```

Reply:
Sorry, I don't see the diff, is it a space? Github doesn't show this properly.

Reply:
It's "only holder **or** spender" not "only holder of spender"

Reply:
thank you

Reply:
Fixed in https://github.com/morpho-org/morpho-token-upgradeable/pull/91/commits/6996f996e3e134e3ddf9e3f71c61fb25de2d5900.